Fix default variables

TheLimeGlass opened the pull request:
### Description
Fixes default variables not working for any expression that wasn't an object.
Issue explained in https://github.com/SkriptLang/Skript/issues/2174
Also added automatic delete of the default variable template when the script that it was defined in removes the variable from `variables:` custom event. https://github.com/SkriptLang/Skript/issues/4567

I have also added support for multiple default variables: `{example::%entity%::%object%}`
I also fixed the super classes lookup, so for example if the expression in the default variable was an entity, and during runtime a pig was provided, it would error because the first super class it got was livingentity and not entity, so it would also error then.

---
**Target Minecraft Versions:** any
**Requirements:** none
**Related Issues:** https://github.com/SkriptLang/Skript/issues/2174 https://github.com/SkriptLang/Skript/issues/4567

TheLimeGlass commented:
Default variable templates don't get deleted. I'm going to fix that aswell in this pull request. Details at https://github.com/SkriptLang/Skript/issues/4567

Work in progress.

TheLimeGlass commented:
Added fixes for https://github.com/SkriptLang/Skript/issues/4567

APickledWalrus reviewed (approved):
This looks okay to me. I think a second review from someone else would be beneficial though.

Also, would there be any way for us to create a test for this?

TheLimeGlass commented:
When I get time to reset to Java 17 i'll add tests.

TheLimeGlass commented:
Currently in draft. Somehow this is now making things weird.

TheLimeGlass commented:
I have also added support for multiple default variables: {example::%entity%::%object%}
I also fixed the super classes lookup, so for example if the expression in the default variable was an entity, and during runtime a pig was provided, it would error because the first super class it got was livingentity and not entity, so it would also error then.

This has turned into a pretty big addition and bug fix.

TPGamesNL commented on src/main/java/ch/njol/skript/ScriptLoader.java:
```diff
@@ -191,6 +196,8 @@ public String toString() {
 	 */
 	private static final Map<String, Set<String>> commandNames = new HashMap<>();
+
+	private static final Multimap<String, String> defaultVariables = HashMultimap.create();
```
Add a Javadoc comment to this, e.g. saying what key and value means

TPGamesNL commented on src/main/java/ch/njol/skript/lang/Variable.java:
```diff
@@ -294,14 +295,20 @@ public <R> Variable<R> getConvertedExpression(Class<R>... to) {
+		Object value = !list ? convertIfOldPlayer(name, event, Variables.getVariable(name, event, local)) : Variables.getVariable(name, event, local);
+		// Check for default variables.
+		if (value == null) {
+			for (String typeHint : this.name.getDefaultVariableNames(event)) {
+				value = Variables.getVariable((local ? LOCAL_VARIABLE_TOKEN : "") + typeHint, event, local);
```
This will make variable getting much slower, as an expression like `{something::%player%}` can result in 9 getVariable calls (if the var isn't set and there isn't a default variable for it):
![image](https://user-images.githubusercontent.com/29547183/175820868-f97d0409-d0b1-46be-8558-33db28c6e64a.png)
and something like `{relation::%player%::%{_p}%}` in 65 getVariable calls, it goes up exponentially

This system should be more efficient, I don't think such a performance cost is worth it.

TheLimeGlass replied:
Yes I did take this into consideration, but it's already doing this now. I had a thought to somehow have a scriptinfo reference and check if the script had default variables, but that's alot of implementation. Do you have an alternative solution?

TheLimeGlass commented on src/main/java/ch/njol/skript/lang/Variable.java:
```diff
@@ -294,14 +295,20 @@ public <R> Variable<R> getConvertedExpression(Class<R>... to) {
-			return Variables.getVariable((local ? LOCAL_VARIABLE_TOKEN : "") + name.getDefaultVariableName(), e, false);
-		return val;
+		Object value = !list ? convertIfOldPlayer(name, event, Variables.getVariable(name, event, local)) : Variables.getVariable(name, event, local);
+		// Check for default variables.
```
Like a `isDefaultVariable()` type method? How would you add this to variables? Also I don't see a use for letting the add-on know about default variables. The add-on just needs to value.

TPGamesNL replied:
My original question was wrong, I didn't think right about default variables.

An issue I see is that an addon may have some syntax related to getting a variable. For example, when a script has default default variable `{stuff::%player%}`, and an addon tries to get the variable `{stuff::TPGamesNL}`, this will not use the default value if needed, because `Variables#getVariable` takes a string for the variable name, without context of the types of parts of the string. 

We should probably add a note to the Javadocs of `Variables` that it is preferred to use methods from `Variable` when using variables from user input in scripts, because they do take default variables into account.

TheLimeGlass commented on src/main/java/ch/njol/skript/lang/Expression.java:
```diff
@@ -149,16 +149,39 @@ default public Stream<? extends T> stream(final Event e) {
+	 * @return A supertype of the first component returned by {@link #getArray(Event)}
+	 */
+	@SuppressWarnings("unchecked")
+	default public Class<? extends T> getReturnType(Event event) {
```
With the additions of structures and sections, I see a bigger usage in this. Especially considering the default variables is a Structure.

TPGamesNL replied:
I personally don't, I don't think the exact return type without being allowed to get the value is very useful.

PS an interface method doesn't need to be public

TheLimeGlass replied:
Everything's public

TheLimeGlass commented on src/main/java/ch/njol/skript/lang/Variable.java:
```diff
@@ -294,14 +295,20 @@ public <R> Variable<R> getConvertedExpression(Class<R>... to) {
+		Object value = !list ? convertIfOldPlayer(name, event, Variables.getVariable(name, event, local)) : Variables.getVariable(name, event, local);
+		// Check for default variables.
+		if (value == null) {
+			for (String typeHint : this.name.getDefaultVariableNames(event)) {
```
My only solution would be to cache the last known value from the getReturnType(event) inside the Variable object and clear it when it's updated. At least this way it would know it's type guaranteed. Do you have any alternative solution?

TheLimeGlass replied:
Alright, I took over the TypeHints class since it was never used and was a good implementation. I created a logger handler type system that collects the runtime return type and variable names expressions' return types.

TheLimeGlass commented on src/main/java/ch/njol/skript/lang/Variable.java:
```diff
@@ -294,14 +295,20 @@ public <R> Variable<R> getConvertedExpression(Class<R>... to) {
+		// Check for default variables.
+		if (value == null) {
+			for (String typeHint : this.name.getDefaultVariableNames(event)) {
+				value = Variables.getVariable((local ? LOCAL_VARIABLE_TOKEN : "") + typeHint, event, local);
```
I think it's good to have.

TPGamesNL replied:
I feel like local variables shouldn't be affected by anything outside of the trigger, it's why they're local after all.

TheLimeGlass replied:
Fair point

TPGamesNL commented:
> I just found out about #3864 maybe this pull request should re enable them too.

If you can fix the issues mentioned in that issue, sure. But it may be better suited for another PR

TheLimeGlass commented:
Ready for review

TheLimeGlass commented:
This has a cluster of git history fixes needing to be changed, as I was told to make it for 2.6 but it didn't end up making 2.6 so it got reverted back to 2.7 and now it needs git work.

TheLimeGlass commented:
Closing in favour of reset in new pull request at https://github.com/SkriptLang/Skript/pull/5317
